irc/histserv.go

```diff
@@ -0,0 +1,242 @@
+			helpShort: `$bDELETE$b deletes an individual message by its msgid.`,
+			enabled:   histservEnabled,
+			minParams: 1,
+			maxParams: 2,
+		},
+		"export": {
+			handler: histservExportHandler,
+			help: `Syntax: $bEXPORT <account>$b
+
+EXPORT exports all messages sent by an account as JSON. This can be used
+for regulatory compliance, e.g., article 15 of the GDPR.`,
+			helpShort: `$bEXPORT$b exports all messages sent by an account as JSON.`,
+			enabled:   historyComplianceEnabled,
+			capabs:    []string{"history"},
+			minParams: 1,
+			maxParams: 1,
+		},
+		"play": {
+			handler: histservPlayHandler,
+			help: `Syntax: $bPLAY <target> [limit]$b
+
+PLAY plays back history messages, rendering them into direct messages from
+HistServ. 'target' is a channel name (or 'me' for direct messages), and 'limit'
+is a message count or a time duration. Note that message playback may be
+incomplete or degraded, relative to direct playback from /HISTORY or
+CHATHISTORY.`,
+			helpShort: `$bPLAY$b plays back history messages.`,
+			enabled:   histservEnabled,
+			minParams: 1,
+			maxParams: 2,
+		},
+	}
+)
+
+// histNotice sends the client a notice from HistServ
+func histNotice(rb *ResponseBuffer, text string) {
+	rb.Add(nil, histServMask, "NOTICE", rb.target.Nick(), text)
+}
+
+func histservForgetHandler(server *Server, client *Client, command string, params []string, rb *ResponseBuffer) {
+	accountName := server.accounts.AccountToAccountName(params[0])
+	if accountName == "" {
+		histNotice(rb, client.t("Could not look up account name, proceeding anyway"))
+		accountName = params[0]
+	}
+
+	server.ForgetHistory(accountName)
+
+	histNotice(rb, fmt.Sprintf(client.t("Enqueued account %s for message deletion"), accountName))
+}
+
+func histservDeleteHandler(server *Server, client *Client, command string, params []string, rb *ResponseBuffer) {
+	var target, msgid string
+	if len(params) == 1 {
+		msgid = params[0]
+	} else {
+		target, msgid = params[0], params[1]
+	}
+
+	accountName := "*"
+	hasPrivs := client.HasRoleCapabs("history")
+	if !hasPrivs {
+		accountName = client.AccountName()
+		if !(server.Config().History.Retention.AllowIndividualDelete && accountName != "*") {
+			hsNotice(rb, client.t("Insufficient privileges"))
+			return
+		}
+	}
+
+	err := server.DeleteMessage(target, msgid, accountName)
+	if err == nil {
+		hsNotice(rb, client.t("Successfully deleted message"))
+	} else {
+		if hasPrivs {
+			hsNotice(rb, fmt.Sprintf(client.t("Error deleting message: %v"), err))
+		} else {
+			hsNotice(rb, client.t("Could not delete message"))
+		}
+	}
+}
+
+func histservExportHandler(server *Server, client *Client, command string, params []string, rb *ResponseBuffer) {
+	cfAccount, err := CasefoldName(params[0])
+	if err != nil {
+		histNotice(rb, client.t("Invalid account name"))
+		return
+	}
+
+	config := server.Config()
+	filename := fmt.Sprintf("%s@%s.json", cfAccount, time.Now().UTC().Format(IRCv3TimestampFormat))
```

Comment:
hm, is there any way this can be used dodgily? writing to disk based on account name makes me a bit nervous and I'd like to see whether this can be broken in any way